### PR TITLE
Add draft for extended project template

### PR DESCRIPTION
**Pull request by knyghty**
This is a draft draft until I get around to writing the reference implementation. It's mostly copying files so shouldn't take too long, just need the motivation :)

Happy to have early eyeballs on it though.

### CODE REVIEW

**Comment by czue** [comment]
Is this the appropriate place to have open-ended discussion about this topic? Or should that be reserved for the forum?

Big picture here's my two cents:

1. I love the idea of this initiative, though I think there will be widespread disagreement about what exactly should go into a new template.
2. I think this is the most important part of the DEP, "it will set a precedent for adding new templates into Django" and would advocate for a strong focus in making it possible to more easily use different templates. So much so that I wonder whether "single-command startproject templates with 2-3 official reference implementations" might be a more suitable DEP (or at least easier to get consensus on). Yes, I know you can already pass a folder or URL to startproject / startapp and I still think we can do better¸e.g. by having a list of options for people to compare in the docs, or potentially by making it easier for template authors to make them more discoverable and configurable.

**Comment by carltongibson** [comment]
Hiya. Thinking actively on this, rather than not responding. 😅

I think here is a fine place to discuss, at least to get to good draft status. 

The motivation section is key. (… 🤔 TBC)

The current project isn't bad — it's just too many files for a project just kicking off. I think it's as we reach that level of complexity and greater that opinions really start to ramify, so keeping the existing template, and using that as the outer limit of what Django itself will provide is perhaps helpful. That leaves space for a couple of simpler templates. 

Then, pondering the `startproject` command. (… 🤔 TBC)

**Comment by knyghty** [comment]
One thing I thought of, though I think it's a bad idea, is related to Cory's idea (maybe exactly the same), is that to get whatever new thing to be the default without making it the default could be to have a new command, e.g. `django-admin new` that would ask some questions and choose a template based on that.

I'm not sure if it's worth putting in the DEP, because it would either require a package to handle the "UI", a good amount of code I'm unsure how to write, or a very basic interface (such as when choosing options when creating migrations). Leaning towards the last of those three things. And of course having --noinput to use whatever we choose as the new default. Then startproject can stay around as is, or be eventually removed in favour of `new`.

It would change the scope of the DEP though, maybe not in a bad way, but still, wasn't what anyone had in mind I think.

Just random ideas :)

**Comment by knyghty** [comment]
To expand slightly, the question could be as simple as:

What type of project do you want to create?

1. A standard project (default)
2. A fancy pants project with frontend things added for you (???)
3. A single file quickstart project for prototyping
4. A third party app
5. A classic project

Where a classic project would be the current template.

More things to maintain, though. And going this road leads to questions: should we add a pyproject.toml, how much configuration should we have, how opinionated should we be, etc. I would err on the side of doing as little as possible, though `pyproject.toml` feels standard enough to add for the user by now. But I also don't want to get lost in the rabbit hole of `.editorconfig`, `.gitignore` and other things that _I_ always use, but are probably too opinionated (some people are probably still using mercurial, for example).

**Comment by knyghty** [comment]
> The motivation section is key. (… 🤔 TBC)

I agree here and I think my motivation section is too light. It doesn't feel very compelling to me. Which is strange because a lot of people have strong opinions on why we need to change it, including myself. Open to ideas, but I also plan to properly sit with this and think it over.

**Comment by carltongibson** [comment]
> … properly sit…

Yes, me too. That’s why just the initial comment. 🤹

**Comment by czue** [comment]
Agree the motivation is key. If the motivation is solely as-written (improving the tutorial experience) then my comment/position doesn't really apply. My motivation is improving the speed and choices for people who want build things with Django.

> wasn't what anyone had in mind I think.

Who is the "anyone" referenced here? When i read through the linked forum posts I don't really see any obvious consensus, so just wondering if there's a backchannel motivation for this specific change (or a discussion I missed elsewhere). I don't want to derail an initiative to make progress because it's not *enough* progress if I'm bringing in requirements that feel out-of-scope.

If we *are* trying to broaden the scope I would have more to say (and think about!) but will hold off for the time being...

**Comment by knyghty** [comment]
> Who is the "anyone" referenced here

Anyone. Nothing backchannel. But I didn't see this mentioned anywhere. So we have a consensus (I think) to do something, but the approaches vary, and the one I just mentioned hasn't been talked about that I know of.

But this broadening of the scope would allow us to do more (possibly: enough) without breaking backwards compatibility. But I do need to go think about it.

**Comment by Kagee** [comment]
I really like this suggestion. To not do to much in one, it sounds sensible to only add the required templates and files, and not cram in other mive-to-have files. A project.toml may be just simple enough though.

Since it will also potentially draw in many new dependencies, maybe the fancy-pants option should be dropped, and replaced with a (as last option) "Paste a link to a project template" as startproject supports now, but is somewhat hidden?

**Comment by knyghty** [comment]
Since there was a look of positive feedback on the ``new`` command I have included this in the proposal.

I've also fleshed out the motivation section for different user types.

I will try to write up the reference implementation during this weekend.

**Comment by Kagee** [comment]
I suggest making it clear*er* in the text/wording that the new default project will use the "merged project/core app layout", not just a classic project and a classic preinstalled core app. While it is "explained" using the tree view, i feel it is not clear from just reading the text.

**Comment by Kagee** [comment]
Have you considered referencing the different templates using slug-names, so they could be chosen when batch-processing, possibly based on the folder-names for the templates to be used? merged-project-app, single-file, third-party-app, classic-project-app?

File: draft/0015-extended-startproject.rst [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+here.
+
+Relevant discussions
+====================
+
+* `Django New Project Structure/Name <https://forum.djangoproject.com/t/django-new-project-structure-name/9987>`_
+* `Updating the default startapp template <https://forum.djangoproject.com/t/updating-the-default-startapp-template/24193>`_
+* `Update startproject with default login/signup/logout options <https://forum.djangoproject.com/t/update-startproject-with-default-login-signup-logout-options/35175>`_
+* `The Single Folder Django Project Layout <https://noumenal.es/notes/django/single-folder-layout/>`_
+
```
Comment (shaib) [inline draft/0015-extended-startproject.rst:201]:
```suggestion
* `The Single Folder Django Project Layout <https://noumenal.es/notes/django/single-folder-layout/>`_
* `Why do we need apps? <https://forum.djangoproject.com/t/why-do-we-need-apps/827/>`_
```

This was a long discussion, my own contribution is post 20 out of (currently) 46. I think it's worth mentioning because the current proposal makes it easier to promote the "single app for project" model, where instead of trying to separate models etc into apps from the get-go, you start with a single app and carve parts of it out to (reusable) apps once you can identify cohesive components whose coupling to the rest of the project can be loosened.

Comment (knyghty) [reply draft/0015-extended-startproject.rst:201]: Thanks Shai 👍🏻

**Comment by knyghty** [comment]
I added a link to a reference implementation. I didn't test that the project templates actually work, but I think it gets the idea across.

One thing I noticed, `startapp` has a context variable `camel_case_app_name` used for the `AppConfig` class name. We'd likely need to add something like `camel_case_proejct_name` to `startproject` as well, to get this working there.

**Comment by Kagee** [comment]
Afaik camel_case_project_name already works? Ref myself https://github.com/Kagee/django-single-folder-project-template/blob/main/project_name/apps.py-tpl#L4

**Comment by knyghty** [comment]
Indeed. Not very well documented :)

I've changed that now in the implementation. Thanks for the heads up.

**Comment by knyghty** [comment]
@czue I'm not planning this as a replacement for startproject, which already does this. You can still use a tempalte URL by pasting it instead of a number, this is mentioned in the prompt of the reference implementation when you use it: https://github.com/knyghty/django-new/blob/main/django_new/management/commands/new.py#L26

**Comment by knyghty** [comment]
@carltongibson, since I have you down as shepherd - I've advertised this everywhere I can think of and only gotten positive responses and mostly crickets here.

What do you think? Should we merge? Do I need to drum up support? Wait to see how the SC organises itself? Something else?

**Comment by carltongibson** [comment]
@knyghty I think we should let the SC election unfold, and circle back to this in January. (We can't advance it without a functioning SC in any case.)

**Comment by hvdklauw** [comment]
Just throwing one thing in there, should the command be called new or init?

Just from the top of my head:
- git init
- uv init

**Comment by shaib** [comment]
FWIW `cargo` has both `new` and `init` -- where `new` is the one in normal use, and `init` is for creating a new package in an existing directory (in line with `git init`). `poetry` also has `new` and `init` behaving similarly (`new` requires a directory path).

Current `startproject` expects a path for an existing directory (or uses the current), so it's more like these `init`s; I'm not quite sure which of these should be taken as a guide, but I'm noting that the goal is to be more accessible and predictable to people new to Django, and unlike `startproject`, both `poetry` and `cargo` had their CLI designed in the last decade.

**Comment by carltongibson** [comment]
Hey @knyghty -- I'd like to progress this during the current cycle. I'll raise it with the SC soon(ish). If you want to IDK go over anything please do, otherwise I'll have a proper read again and feedback soon(ish). 

(Sorry for the ishes 😜)

**Comment by knyghty** [comment]
The DEP itself is good to review. The only remaining TODOs (and perhaps aren't needed before judging this) are for the linked reference implementation:
- Add a pyproejct.toml to the "default" proejct
- Add a template for a third party app

I do hope to get to these soon, but not sure if they're necessary to have fully in place or not (suspect not).

**Comment by luzfcb** [comment]
My comment may not be directly related to the DEP and out of scope, but I'd like to mention it because implementing this DEP maybe could impact the code currently used by `startapp` / `startproject`.

In [cookiecutter-django](https://github.com/cookiecutter/cookiecutter-django) (If you haven't heard of it, it's basically a highly opinionated django project template), we place all our Django apps within a namespace. This approach keeps the project better organized in the long run (in our opinion) and simplifies the process of packaging and distributing the apps (on our internal pypi like server)

Therefore, when creating a new app in a project based on the cookiecutter-django template, we need to:

1 - create the `<name-of-the-app>` app with `python manage.py startapp`
2 - move `<name-of-the-app>` directory to `<project_slug>` directory
3 - edit `<project_slug>/<name-of-the-app>/apps.py` and 
change `name = "<name-of-the-app>"` to `name = "<project_slug>.<name-of-the-app>"`
4 - add `"<project_slug>.<name-of-the-app>.apps.<NameOfTheAppConfigClass>",` on your [`LOCAL_APPS` on `config/settings/base.py`](https://github.com/cookiecutter/cookiecutter-django/blob/7c6dff3b707393d49c20f1639da048272fd01b81/%7B%7Bcookiecutter.project_slug%7D%7D/config/settings/base.py#L108)

In the past, I tried to extend the `startapp` code to create a new `startapp` like management command that supported creating the app in a namespace, but at that time, most of the `startapp` logic was contained within a single method and this would have been difficult to maintain in the long run.

In short, I basically wanted to run

```
django-admin startapp namespace.app_name
```

and have django create the directory structure and app respecting the namespace, that is:

```
namespace/
    __init__.py
    app_name/
        __init__.py
        admin.py
        apps.py
        models.py
        migrations/
               __init__.py
        tests.py
        views.py
```
and if the namespace already exists but the app_name not, simple create the app_name inside the namespace

The question is: Are there any plans to make `startproject` code and/or `startapp` code a bit more reusable and extensible by third parties?

Or perhaps simply implement namespace support in `startapp`?

**Comment by knyghty** [comment]
> Or perhaps simply implement namespace support in startapp?

This part could arguably be in scope, because the new "default" startproject template proposed here comes with (at least) a bit of an implication that nesting apps under some global `project_name` namespace is a good idea, and I personally am a proponent of this.

I haven't given any thought to making `startproject` and `startapp` more extensible, and I would consider it out of scope for this DEP.

File: accepted/0015-extended-startproject.rst [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Of these, this proposal makes 3 and 4 easier, and 5 unnecessary as the default
+structure will be faster to spin up and get working with, and the single file
+project will always work with the current branch.
+
+Third party apps
+----------------
+
+Setting up a third party app can be quite complex. Something as simple as
+running tests or the ``makemigrations`` command can be quite tricky to figure
+out.
```
Comment (LilyFirefly) [inline accepted/0015-extended-startproject.rst:162]: I would like to see an example of this in the demo repo - currently it's not clear how this would differ from the other templates. Alternatively, more detail could be added to the DEP.

Comment (tim-schilling) [inline accepted/0015-extended-startproject.rst:162]: I'm not sure I understand the point this section is trying to make. It's claiming there's complexity, but it's not clear what that complexity is, why it exists and how this would improve it. Can you clarify that for me please? I'm not sure any part of the DEP needs to change though. I want to understand to start.

File: draft/0015-extended-startproject.rst [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+1. Using a project using models from the test suite.
+2. The user maintains their own test project that expands over time.
+3. Spinning up a fresh project each time.
+4. Using a ready built site such as ``django-admind-demo``.
+5. Using one of the various single file templates from the web.
+
+Of these, this proposal makes 3 and 4 easier, and 5 unnecessary as the default
+structure will be faster to spin up and get working with, and the single file
+project will always work with the current branch.
```
Comment (tim-schilling) [inline draft/0015-extended-startproject.rst]:
```suggestion
Of these, this proposal makes 3 and 4 easier, and 5 unnecessary as the single file
project will always work with the current branch.
```

I'm not sure the point about the default template structure applies to a concern about using a custom single file template.

File: accepted/0015-extended-startproject.rst (hunk 2 above)
Comment (knyghty) [reply accepted/0015-extended-startproject.rst:162]: There are two points of complexity listed.

There isn't really an obvious way to set your third party package up to run tests. I'm not sure if we should be so opinionated here since a lot of packages will want to use pytest. But adding something as a start could be useful. And it's the same thing, if your package has migrations, generating those migrations can be annoying, as you essentially need to plug it into a Django project to create the migration files.

Another thing is having a project set up under tests so you can run the tests, you will usually need to set up a secret key and databases, etc.

There is currently not really any documented way to do any of these, and a project template could help here.

File: draft/0015-extended-startproject.rst (hunk 3 above)
Comment (knyghty) [reply draft/0015-extended-startproject.rst]: Yes, the waffle applies to 3 and 4 but it's already clear. I like this better.

File: accepted/0015-extended-startproject.rst (hunk 2 above)
Comment (tim-schilling) [reply accepted/0015-extended-startproject.rst:162]: Ahhhh. I see it now. Thank you for the clarification! Good to know for Django Commons too. What do you think about adding the following sentence at the end? "Because these operations require a Django project, they are improved by making it easy to generate a project." I think it clarifies exactly what you just conveyed and helps other people who aren't familiar with creating third-party packages understand the benefit a bit more.

Comment (knyghty) [reply accepted/0015-extended-startproject.rst:162]: What I'm trying to get across here is the utility in having a template for third party apps included. I'm not suggesting you generate a project to make it easier to write your third party app, rather that we can generate the app structure entirely, including the embedded project. So maybe I need to be clearer about that, but not sure how.

Comment (tim-schilling) [reply accepted/0015-extended-startproject.rst:162]: What I understood is if I have a third party package that needs to have a migration, I need to create a separate Django project to generate the migration. My assumption is that having a template to generate a whole django project would make it easier for me to do that work to create a migration to include in my third party package. That django project would be discarded after creating the migration file.

Comment (knyghty) [reply accepted/0015-extended-startproject.rst:162]: Ah that part is reasonable. I think this will become more clear with a third party app template, but I think there's a good amount of bikeshedding possible there so I was putting it off a bit.

Comment (tim-schilling) [reply accepted/0015-extended-startproject.rst:162]: Fair enough. I'm good leaving this as is for now. I don't think I hold this opinion strongly enough to ask you to change it.

File: accepted/0015-extended-startproject.rst [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+3. A template suitable for writing a third party app.
+4. A "classic" project structure using the current ``startproject`` template.
+
+
+A URL can also be entered that works the same way as ``startproject``.
+
+If ``--noinput`` is passed to the command, it will select the default template.
+
+This command could be extended later either by contributions to Django or perhaps
+by third party packages.
```
Comment (LilyFirefly) [inline accepted/0015-extended-startproject.rst:119]: I would really strongly like support for pluggable third party packages to be added in the initial implementation. This will make it really easy to experiment with new options and reduce the pressure to add more to core before they've proven themselves in the third-party ecosystem.

**Review by LilyFirefly — APPROVED** [review]

File: draft/0015-extended-startproject.rst [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+Contributors often need to set up a new project to test their changes. It's
+not really documented anywhere how to do this, so there is no standard process.
+This can be a barrier to entry for new contributors. Examples of current
+methods include:
+
+1. Using a project using models from the test suite.
+2. The user maintains their own test project that expands over time.
+3. Spinning up a fresh project each time.
+4. Using a ready built site such as ``django-admind-demo``.
```
Comment (tim-schilling) [inline draft/0015-extended-startproject.rst]:
```suggestion
4. Using a ready built site such as ``django-admin-demo``.
```

File: accepted/0015-extended-startproject.rst (hunk 4 above)
Comment (knyghty) [reply accepted/0015-extended-startproject.rst:119]: Yeah that's part of it, this text is meant to show it should be easy, but I agree the wording here is probably not strong enough. I will add something here.

Comment (knyghty) [reply accepted/0015-extended-startproject.rst:119]: I am apparently sleep-deprived and talking nonsense in a now forever-deleted message :)

But the main point remains: I am not sure how to implement this in a way that's really pluggable since at this point we have nothing to start from.

**Review by tim-schilling — APPROVED** [review]

File: accepted/0015-extended-startproject.rst (hunk 4 above)
Comment (LilyFirefly) [reply accepted/0015-extended-startproject.rst:119]: An idea that came up in today's Steering Council meeting would be to add a dependency on [`pluggy`](https://github.com/pytest-dev/pluggy) and write the `django new` command in a way that allows:

```
$ pip install django
$ pip install my-plugin
$ django new
  1: A standard project (default)
  2: A fancy pants project with frontend things added for you (???)
  3: A single file quickstart project for prototyping
  4: A third party app
  5: A classic project
  6: My plugin
```

This particular idea isn't *decided* yet - we'd like to check in with the community first that this is a good idea. It would also unlock using `pluggy` in future features if we decided that was a good idea.

Comment (LilyFirefly) [reply accepted/0015-extended-startproject.rst:119]: I've created a new-features post about adding `pluggy` to Django: https://github.com/django/new-features/issues/88

**Comment by LilyFirefly** [comment]
The Steering Council has discussed this DEP and we're happy to approve adding the `django new` command and approve in principle the various proposed templates, which can be added individually as and when they're ready.

File: draft/0015-extended-startproject.rst [hunk 6]
```diff
@@ -0,0 +1,211 @@
+=======================================
+DEP 15: Improved startproject interface
+=======================================
+
+:DEP: 15
+:Author: Tom Carrick
+:Implementation Team: Tom Carrick
+:Shepherd: Carlton Gibson
+:Status: Draft
```
Comment (LilyFirefly) [inline draft/0015-extended-startproject.rst]:
```suggestion
:Status: Accepted
```

File: accepted/0015-extended-startproject.rst
Comment (LilyFirefly) [inline accepted/0015-extended-startproject.rst:1]: Move this to `accepted/0015-extended-startproject.rst`.

**Review by LilyFirefly — CHANGES_REQUESTED** [review]
